### PR TITLE
Docker image with precompiled Pluto sysimage

### PR DESCRIPTION
**Pull request by lungben**
Hi,

I tried to make a Docker image with a precompiled Pluto (+ PlutoUI and Plots) sysimage for faster startup / first plots.
You can check it out here:
`docker run -p 1234:1234 plutojl/pluto:precompiled`

While it indeed starts faster, it has also some drawbacks to the standard Pluto image:

- Larger image size: 1.6 GB compared to 580 MB
- Building the image on Dockerhub does not work - when running an image built there I get the error message "ERROR: Unable to find compatible target in system image.". Maybe the Docker build machine uses a different architecture? However, building local on my machine works, at least when I run it on the same machine (Win10, 64-bit x86).

### CODE REVIEW

**Comment by fonsp** [comment]
Awesome awesome! @shashankp @dralletje

**Comment by fonsp** [comment]
You used the tests as warmup right? Maybe you can try this: https://github.com/fonsp/pluto-on-binder/blob/6d9c3db0caab93797368988ae2e08d8a3e0d13a0/warmup.jl - this also includes precompilation of HTTP.jl.

How much of that 1.6GB is Plots.jl?

**Comment by lungben** [comment]
I tried it with the Pluto tests, but it caused build failures on Dockerhub (maybe due to timeouts?). Your warmup script looks great, I'll try this out.
Regarding Plots, I can remove it to check how this influences image size. An alternative would be to include it only as a precompiled package, but not in the sysimage.

What still puzzles me is the lack of portability of the sysimages. I could not start the image on my Windows PC which was built on Dockerhub and my DiskStation could neither start the Dockerhub nor the Windows image. I thought that a Docker image would run on all host systems the same way.

**Comment by Roger-luo** [comment]
maybe we can consider put this package to docker directly since I have built a set of tools let you compile system image in CI in Comonicon for the CLI and automatically download it when a system image is available.

**Comment by Roger-luo** [comment]
I can enable the system image build in this package on github action if it's needed. But I'll need @fonsp to setup a ssh-key for tagbot

**Comment by fonsp** [comment]
Sure just tell me what to type where

**Comment by shashankp** [comment]
> Hi,
> 
> I tried to make a Docker image with a precompiled Pluto (+ PlutoUI and Plots) sysimage for faster startup / first plots.
> You can check it out here:
> `docker run -p 1234:1234 plutojl/pluto:precompiled`
> 
> While it indeed starts faster, it has also some drawbacks to the standard Pluto image:
> 
> * Larger image size: 1.6 GB compared to 580 MB
> * Building the image on Dockerhub does not work - when running an image built there I get the error message "ERROR: Unable to find compatible target in system image.". Maybe the Docker build machine uses a different architecture? However, building local on my machine works, at least when I run it on the same machine (Win10, 64-bit x86).

I see the same issue, builds & runs on Win10,
but fails on macOS with SIGKILL(9) 

Wondering whether Plots is *required* to get the docker image up & running?
If currently precompiling it is an issue, why not leave it out now. Once a user launches the notebook, he/she can include it.

Also [this change](https://github.com/fonsp/Pluto.jl/commit/d0ff39e6e406ba24ae66f90d95be39db8de3aa3a#diff-470b89018397f6ee8352c6b68e9ed1ec) needs to have an additional check for containers
something like 

```
if isfile(".dockerenv") return false
```

**Comment by Roger-luo** [comment]
@fonsp basically just add a SSH_KEY and update TagBot for it, see also: https://github.com/JuliaRegistries/TagBot#ssh-deploy-keys

**Comment by shashankp** [comment]
> However, I get an error message (but this does not prevent building the image):
> 
> ```
> [ Info: Starting HTTP server
> IOError: could not spawn `xdg-open http://localhost:40404/`: no such file or directory (ENOENT)
> ```


error is cos of launch_browser.
For docker @fonsp advice should work
```
Pluto.run("0.0.0.0", 1234; configuration=ServerConfiguration(launch_browser=false))
```

I've tried [multi-stage docker build ](https://docs.docker.com/develop/develop-images/multistage-build/) which brought the image size to ~1GB 
(installed Pluto & Plots, but only Pluto is pre-compiled)

Plots compilation using PackageCompiler is failing & is reproducible; probably opening an issue with them can help.

**Comment by lungben** [comment]
> I've tried [multi-stage docker build ](https://docs.docker.com/develop/develop-images/multistage-build/) which brought the image size to ~1GB
> (installed Pluto & Plots, but only Pluto is pre-compiled)

This is a good idea!
Could you please share your Dockerfile? I have not worked with multi-stage builds yet.

**Comment by shashankp** [comment]
I'm experimenting on a [fork](https://github.com/shashankp/PlutoUtils.jl)
the artifact is on [dockerhub](https://hub.docker.com/r/shashankpv/plutojl/tags); once its in a good shape, can move to official location

**Comment by shashankp** [comment]
+Adding a [Dockerfile](https://raw.githubusercontent.com/shashankp/18S191/master/Dockerfile) in a repo makes it ready for Binder
[https://mybinder.org/v2/gh/shashankp/18S191/master?urlpath=pluto](https://mybinder.org/v2/gh/shashankp/18S191/master?urlpath=pluto)

-Permission issues to be sorted out
-Its still quite slow ~ 2min
-Container image is ~1.2GB even without precompiled Plots

**Comment by fonsp** [comment]
@lungben Why is Clang in the image?

**Comment by fonsp** [comment]
@shashankp I tested your binder and it's fast! 20 sec to main menu, but I couldn't test the time to first notebook because you need to disable URL security:
```julia
Pluto.run("0.0.0.0", 1234;
	configuration=Pluto.ServerConfiguration(launch_browser=false),
	security=Pluto.ServerSecurity(false),
	)
```

**Comment by lungben** [comment]
> @lungben Why is Clang in the image?

The PackageCompiler needs a C compiler, either gcc or Clang. 
I tried both, they roughly yield the same image size. Unfortunately the issue that container built on Windows are not running on Linux, etc. is also present for both compiler.

**Comment by fonsp** [comment]
Aha! So the precompilation step needs a c compiler, but does it need to be in the final image?

**Comment by shashankp** [comment]
nope, multistage build solves that issue.
it uses an image with clang for compiling, then copies over the artifacts to another image which would be the final one.

we're not creating multiple outputs, there's still only one

**Comment by shashankp** [comment]
> > the issue that container built on Windows are not running on Linux, etc.
> That's tough - this means that we would need to create multiple docker images, right? It seems like Julia does this too

Would beat the whole purpose of using containerized environments ☹️
Generally compiling with the right target architecture flags should do the trick.

> ```julia
> Pluto.run("0.0.0.0", 1234;
> 	configuration=Pluto.ServerConfiguration(launch_browser=false),
> 	security=Pluto.ServerSecurity(false),
> 	)
> ```
> 
> Time to first notebook should be where we see the benefit of precompilation.
> 
> As for why it's still 20 seconds and not less - maybe because of the increased image size? I also found some randomness is binder launch times, but I only timed it once.

Fixed the permissions issue & added this setting, but not much difference in timing
I wouldn't worry too much about breaking the 20sec barrier, the notebooks itself should 🚀.
UI test to time a sample notebook should be a good metric.

**Comment by lungben** [comment]
I opened an Issue at PackageCompiler for the error when running on an other host platform, hope that this could be solved.
The multi-stage approach of @shashankp is great, I suggest to use his version for the official image when the cross-platform issue is solved out.

**Comment by lungben** [comment]
When adding the keyword argument `cpu_target = PackageCompiler.default_app_cpu_target()` the image built on Dockerhub works for Windows and Linux (Synology DS).
The image `plutojl/pluto:precompiled` is updated, you could try if it also works for you.

File: docker/precompiled/Dockerfile [hunk 1]
```diff
@@ -21,7 +21,7 @@ RUN julia -e 'using Pkg; \
     Pkg.add(["Pluto", "PlutoUI", "Plots", "PackageCompiler"]); \
 	Pkg.precompile(); \
     using PackageCompiler; \
-    create_sysimage([:Pluto, :PlutoUI]; precompile_execution_file="warmup.jl", replace_default=true)' \
+    create_sysimage([:Pluto, :PlutoUI]; precompile_execution_file="warmup.jl", replace_default=true, cpu_target = PackageCompiler.default_app_cpu_target())' \
```
Comment (shashankp) [inline docker/precompiled/Dockerfile:24]: Include Plots while creating the sysimage; it was the one creating the issue

Comment (lungben) [reply docker/precompiled/Dockerfile:24]: For me, the portability issue was also present after removing Plots. But I'll check if it still works after re-adding Plots.

**Comment by lungben** [comment]
What do we want to have in the sysimage of the official Pluto Docker image (despite Pluto and PlutoUI)?
Especially, should we include Plots.jl?

**Comment by shashankp** [comment]
> What do we want to have in the sysimage of the official Pluto Docker image (despite Pluto and PlutoUI)?
> Especially, should we include Plots.jl?

the current images I've built are only Pluto, only when I've tried adding Plots it failed; 
if that is the case, we're good to go @fonsp

**Comment by fonsp** [comment]
Cool! Why not make the precompiled Docker the default? Maybe with a not_precompiled alternative?

**Comment by fonsp** [comment]
And what is the size of the final image? Did we manage to remove Clang?

> I wouldn't worry too much about breaking the 20sec barrier, the notebooks itself should 🚀.

Exactly - the first 10-20 seconds seem like a given (on binder), but if I open a `/open?url=...` link on binder (like https://pluto-on-binder.glitch.me/), it currently takes 40 seconds. Those additional 20 seconds are because Pluto is compiling, and this is what precompilation should avoid.

**Comment by fonsp** [comment]
Did anyone try to create a binder from our docker image? It's a bit messy because you still need to install python, jupyter, the "pluto extension" for jupyter, like I do here: https://github.com/fonsp/pluto-on-binder We can talk about this in a new issue if you prefer

**Comment by shashankp** [comment]
in <repo-path>/docker/binder
there is another Dockerfile which takes the docker image with precompiled Pluto & adds python+jupyter+creates an extension. I'm calling it [plutojl-binder](https://hub.docker.com/repository/docker/shashankpv/plutojl-binder)

Referring this image is enough to make it work with Binder
[Example](https://github.com/shashankp/18S191/blob/master/Dockerfile)

**Comment by lungben** [comment]
I merged the changes of @shashankp and myself.
As far as I can see, both the non-precompiled and precompiled Docker images are working (at least on Windows, I'll check Linux later).
With Binder and Github Actions I do not have experience yet, I changed them to my best knowledge, but have not tested it yet.

@fonsp I get the following error message in your warmup script, but it compiles nevertheless. Is this an issue?
```
[ Info: Starting HTTP server
IOError: could not spawn `xdg-open http://localhost:40404/`: no such file or directory (ENOENT)
Stacktrace:
 [1] _spawn_primitive(::String, ::Cmd, ::Array{Any,1}) at ./process.jl:99
 [2] #585 at ./process.jl:112 [inlined]
 [3] setup_stdios(::Base.var"#585#586"{Cmd}, ::Array{Any,1}) at ./process.jl:196
 [4] _spawn at ./process.jl:111 [inlined]
 [5] run(::Cmd; wait::Bool) at ./process.jl:439
 [6] run at ./process.jl:438 [inlined]
 [7] open_in_default_browser(::String) at /home/pluto/.julia/packages/Pluto/upwxH/src/webserver/WebServer.jl:42
 [8] run(::String, ::Int64; configuration::Pluto.ServerConfiguration, session::Pluto.ServerSession, security::Pluto.ServerSecurity) at /home/pluto/.julia/packages/Pluto/upwxH/src/webserver/WebServer.jl:201
 [9] #run#209 at /home/pluto/.julia/packages/Pluto/upwxH/src/webserver/WebServer.jl:238 [inlined]
 [10] (::var"#1#2")() at ./task.jl:356[ Info: Warmup done
[ Info: PackageCompiler: creating system image object file, this might take a while...
```

**Comment by lungben** [comment]
Docker repo for binder created:
https://hub.docker.com/r/plutojl/pluto-binder

The 3 images (Pluto default, Pluto precompiled, Pluto-binder) are currently building at Dockerhub (still triggered manually).

**Comment by fonsp** [comment]
Fantastic! Thanks a lot!

**Comment by fonsp** [comment]
@dralletje

**Comment by lungben** [comment]
I am merging this PR now.
Currently, the build of updated Docker images need to be started manually by @fonsp or myself in Dockerhub. This should be replaced by Github Actions in the future, but I would like to track that in a separate issue.

**Comment by lungben** [comment]
https://github.com/fonsp/PlutoUtils.jl/issues/20

**Pull request merged**